### ISSUE TITLE
v2.0-devel branch: mismatch in mapfile / file names

### ISSUE THREAD

**Issue by shulevski** [lofar-astron/prefactor#73]
While running the Initial-Subtract-Deep.parset, I got an error:

> 2016-08-21 02:09:44 DEBUG   genericpipeline.executable_args: Writing data map file: /home/shulevski/P198_57/prefactor/rundir/Initial-Subtract-Deep/mapfiles/wsclean_high1_deep.mapfile
2016-08-21 02:09:44 DEBUG   genericpipeline.executable_args: Writing data map file: /home/shulevski/P198_57/prefactor/rundir/Initial-Subtract-Deep/mapfiles/wsclean_high1_deep-MFS-image.fits.mapfile
2016-08-21 02:09:44 DEBUG   genericpipeline.executable_args: Writing data map file: /home/shulevski/P198_57/prefactor/rundir/Initial-Subtract-Deep/mapfiles/wsclean_high1_deep-MFS-model.fits.mapfile
2016-08-21 02:09:44 INFO    genericpipeline.executable_args: recipe executable_args completed
2016-08-21 02:09:44 ERROR   genericpipeline: *******************************************
2016-08-21 02:09:44 ERROR   genericpipeline: Failed pipeline run: Initial-Subtract-Deep
2016-08-21 02:09:44 ERROR   genericpipeline: Detailed exception information:
2016-08-21 02:09:44 ERROR   genericpipeline: <type 'exceptions.KeyError'>
2016-08-21 02:09:44 ERROR   genericpipeline: 'mapfile'
2016-08-21 02:09:44 ERROR   genericpipeline: *******************************************

The `wsclean_high1_deep` step completed successfully, I believe that the reason for the error is in one of the two following steps: `createmap_high1_deep_image` or `mask_high_deep`.

In the `createmap_high1_deep_image` step there is a line:

> createmap_high1_deep_image.control.filename          =   wsclean_high1_deep_image.mapfile

but in the mapfiles directory there is no mapfile with that name.

There is, however, a mapfile: `wsclean_high1_deep.mapfile` which contains:

> [{'host': 'localhost', 'file': '/home/shulevski/P198_57/prefactor/workdir/Initial-Subtract-Deep/L524836_SB001_uv_128800BF2t_131MHz.pre-cal.wsclean_high1_deep', 'skip': False}]

which I suppose serves to create two other mapfiles (present in the mapfile folder): `wsclean_high1_deep-MFS-image.fits.mapfile` and `wsclean_high1_deep-MFS-model.fits.mapfile`.

These two mapfiles refer to the image (model) produced. For example, the contents of the `wsclean_high1_deep-MFS-image.fits.mapfile` are:

> [{'host': 'localhost', 'file': '/home/shulevski/P198_57/prefactor/workdir/Initial-Subtract-Deep/L524836_SB001_uv_128800BF2t_131MHz.pre-cal.wsclean_high1_deep-MFS-image.fits', 'skip': False}]

However, the contents of the `/home/shulevski/P198_57/prefactor/workdir/Initial-Subtract-Deep/` directory:

> L524836_SB001_uv_128800BF2t_131MHz.pre-cal.wsclean_high1_deep-dirty.fits           L524836_SB001_uv_128800BF2t_131MHz.pre-cal.wsclean_high1_deep-psf.fits
L524836_SB001_uv_128800BF2t_131MHz.pre-cal.wsclean_high1_deep-first-residual.fits  L524836_SB001_uv_128800BF2t_131MHz.pre-cal.wsclean_high1_deep-residual.fits
L524836_SB001_uv_128800BF2t_131MHz.pre-cal.wsclean_high1_deep-image.fits
L524836_SB001_uv_128800BF2t_131MHz.pre-cal.wsclean_high1_deep-model.fits

do not include files with the names stated in the mapfiles, the `-MFS-` part of the filename is missing.

I can see from the parset that the `wsclean_high1_deep` step contains the line:

> wsclean_high1_deep.control.outputsuffixes            =   [-MFS-image.fits,-MFS-model.fits]

which suggests that these files were meant to be created.

In summary, there is a discrepancy in the name of one mapfile which is not created (or renamed) and there are files with different names which are being referenced.

This issue may be related to #56. Maybe someone has already a fix and is ready for merging? I am working with a version of this branch updated 2 weeks ago from the repo code.

**Comment by shulevski** [comment]
Can this issue actually be caused by the fact that the Initial-Subtract-Deep.parset should not be run with the v2.0-devel branch? If there was no code merge between the v2.0-devel and the deepimage branch, it can well be the case...

**Comment by wllwen007** [comment]
How many bands are you combining? It's curious as it looks like it hasn't actually done an MFS clean.

If I'm not mistaken the deepimage branch has been merged into v2.0-devel.

**Comment by shulevski** [comment]
I am processing 60 SBs, combining 10SBs in one image.
The comment on the readme file in the v2.0-devel branch says that v2.0-devel was merged into the deepimage branch, so maybe I should run the deepimage branch. The issue I see is not just with the images not being generated, but the mapfile names not matching. I will try running the deepimage branch with this parset; if that works, removing the deep image parset from the v2.0-devel branch should be considered.

**Comment by twshimwell** [comment]
How many subbands are you using? If the 

wsclean_high1_deep.argument.channelsout              =   nchansout_clean1

gives it one output channel then the MFS files are not produced and it will crash like this...

**Comment by twshimwell** [comment]
I thought we fixed this by putting in this statement into the magic script:

    if nbands > 4:
        nchansout_clean1 = np.int(nbands/4)
    else:
        nchansout_clean1 = np.int(nbands)

however, I guess if you have nbands = 1 then it will still have the problem.

**Comment by twshimwell** [comment]
but if you have no *MFS* images then it seems like an issue. You are processing all 6x10subbands in one deep image run? Or are you processing 6 deep image runs each of 10sbs?

**Comment by shulevski** [comment]
The fix you mention is in the `InitSubtract_deep_sort_and_compute.py` script I have in my local copy. The prefactor setup I used generated 6 combined MSs, so it took my 60 SBs and combined each 10 SBs in a MS consisting of 10 channels each. I guess that since the subtract takes these 6 combined MSs as input, the imaging is done one by one on each of these MSs. The relevant line in the target parset of prefactor is:

> ! num_SBs_per_group    = 10 # make concatenated measurement-sets with that many subbands

**Comment by twshimwell** [comment]
Ah I see. Opps I guess that is a mistake in the initsubtract_deep_sort_and_compute.py. The bit

    if nbands > 4:
        nchansout_clean1 = np.int(nbands/4)
    else:
        nchansout_clean1 = np.int(nbands)

will return 1 if you have nbands=6 like you have. I simple change could be

    if nbands > 8:
        nchansout_clean1 = np.int(nbands/4)
    else:
        nchansout_clean1 = np.int(nbands)

then you will get 6 output bands and the MFS image.

**Comment by shulevski** [comment]
Indeed. I will make the change in my local copy and restart; let's test this.

**Comment by twshimwell** [comment]
Did this fix work for you?

**Comment by shulevski** [comment]
Yes, the fix works, I was waiting for the subtract run to finish before I close the issue; I expect to do so in a few days.

**Comment by shulevski** [comment]
The deepimage branch subtract has finished; closing the issue. Tim, you can implement the fix we discussed if you think that it's robust enough.
